tests/saw/spec/rfc-handshake.cry

```diff
@@ -0,0 +1,302 @@
+(certificateRequest : MessageType) = 5
+(serverHelloDone    : MessageType) = 6
+(certificateVerify  : MessageType) = 7
+(clientKeyExchange  : MessageType) = 8
+(finished           : MessageType) = 9
+(newSessionTicket   : MessageType) = 10
+(certificateStatus  : MessageType) = 11
+
+// ChangeCipherSpec message types
+(changeCipherSpecMessage : MessageType) = 0
+
+// Data message types
+(applicationData : MessageType) = 0
+(error : MessageType) = 1
+
+// Sender (data flow direction)
+type Sender = [2]
+(server   : Sender) = 0
+(client   : Sender) = 1
+(both     : Sender) = 2
+(noSender : Sender) = 3
+
+// Key exchange algorithms
+type KeyExchange = [3]
+(DH_anon : KeyExchange) = 0
+(DHE_DSS : KeyExchange) = 1
+(DHE_RSA : KeyExchange) = 2
+(RSA     : KeyExchange) = 3
+(DH_DSS  : KeyExchange) = 4
+(DH_RSA  : KeyExchange) = 5
+
+// TLS protocol parameters that determine the transition to take in each state.
+// Some of these fields are set as a configuration to the end-point, while others
+// are determined from the messages send by the counterparty. We model both as
+// parameters since the state machine model reasons about both end-points.
+type Parameters = {keyExchange : KeyExchange
+                   //^ Negotiated key exchange algorithm
+                  ,sessionTicket : Bit
+                   //^ The client had a session ticket
+                  ,renewSessionTicket : Bit
+                   //^ Whether the server decides to renew a session ticket
+                  ,sendCertificateStatus : Bit
+                   //^ Whether the server decides to send the certificate status
+                   //  message
+                  ,requestClientCert : Bit
+                   //^ Whether the server requests a certificate from the client
+                  ,includeSessionTicket : Bit
+                   //^ Whether the server includes a session ticket extension in
+                   //  SERVER_HELLO
+                  }
+
+// A relation between the s2n_connection struct and the parameters
+connectionParameters : connection -> Parameters -> Bit
+connectionParameters conn params =
+    conn.server_can_send_ocsp == params.sendCertificateStatus
+ /\ ((conn.key_exchange_eph /\ keyExchangeEphemeral params) \/
+     (~conn.key_exchange_eph /\ keyExchangeNonEphemeral params))
+ /\ (conn.is_caching_enabled /\ ~conn.resume_from_cache) == params.sessionTicket
+ /\ (~params.includeSessionTicket) // s2n server does not issue tickets at this time
+ /\ (~params.renewSessionTicket) // s2n server does not issue tickets at this time
+ /\ conn.client_auth_flag == params.requestClientCert
+
+// A predicate that tells whether key-exchange is non-ephemeral (doesn't use
+// Diffie-Hellman key exchange). In this case the key exchange algorithm does not
+// require KeyExchange messages.
+keyExchangeNonEphemeral : Parameters -> Bit
+keyExchangeNonEphemeral params = params.keyExchange == RSA \/ params.keyExchange == DH_DSS \/ params.keyExchange == DH_RSA // 
```

Comment:
It would be possible if there wasn't `DH_anon`, which `s2n` does not support (for good reason, of course). (Both `keyExchangeEphemeral` and `keyExchangeNonEphemeral` are false for that key exchange method.)

We actually only use `keyExchangeNonEphemeral` in the RFC-derived state machine, and use `keyExchangeEphemeral` only when relating RFC-level parameters to the abstracted `connection` struct (https://github.com/GaloisInc/s2n/blob/master/tests/saw/spec/rfc-handshake.cry#L121). It was actually previously implemented like the rewrite you are proposing, and it caused a proof error because it admitted `DH_anon` key exchange in the `connectionParameters` relation and the RFC and s2n  state machine paths diverge due to that. 

However, we could reduce the complexity by dropping `keyExchangeEphemeral` and rewrite `connectionParameters` relation to only use `keyExchangeNonEphemeral` and explicitly preclude `DH_anon`. E.g., something like the following

```Cryptol
// A relation between the s2n_connection struct and the parameters
connectionParameters : connection -> Parameters -> Bit
connectionParameters conn params =
    conn.server_can_send_ocsp == params.sendCertificateStatus
 /\ ((conn.key_exchange_eph /\ ~keyExchangeNonEphemeral params /\ params.keyExchange != DH_anon) \/
     (~conn.key_exchange_eph /\ keyExchangeNonEphemeral params))
 /\ (conn.is_caching_enabled /\ ~conn.resume_from_cache) == params.sessionTicket
 /\ (~params.includeSessionTicket) // s2n server does not issue tickets at this time
 /\ (~params.renewSessionTicket) // s2n server does not issue tickets at this time
 /\ conn.client_auth_flag == params.requestClientCert
```

I'll investigate whether that works and get back to you.

Alternatively, we could change the comments to better reflect the intended semantics.